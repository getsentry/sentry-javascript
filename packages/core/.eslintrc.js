module.exports = {
  extends: ['../../.eslintrc.js'],
  ignorePatterns: ['rollup.npm.config.mjs', 'test/utils-hoist/buildPolyfills/originals.js'],
};
