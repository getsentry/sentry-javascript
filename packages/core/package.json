{
  "name": "@sentry/core",
  "version": "7.120.4",
  "description": "Base implementation for all Sentry JavaScript SDKs",
  "repository": "git://github.com/getsentry/sentry-javascript.git",
  "homepage": "https://github.com/getsentry/sentry-javascript/tree/master/packages/core",
  "author": "Sentry",
  "license": "MIT",
  "engines": {
    "node": ">=8"
  },
  "files": [
    "cjs",
    "esm",
    "types",
    "types-ts3.8"
  ],
  "main": "build/cjs/index.js",
  "module": "build/esm/index.js",
  "types": "build/types/index.d.ts",
  "typesVersions": {
    "<4.9": {
      "build/types/index.d.ts": [
        "build/types-ts3.8/index.d.ts"
      ]
    }
  },
  "publishConfig": {
    "access": "public",
    "tag": "v7"
  },
  "dependencies": {
    "@sentry/types": "7.120.4",
    "@sentry/utils": "7.120.4"
  },
  "scripts": {
    "build": "run-p build:transpile build:types",
    "build:dev": "yarn build",
    "build:transpile": "rollup -c rollup.npm.config.mjs",
    "build:types": "run-s build:types:core build:types:downlevel",
    "build:types:core": "tsc -p tsconfig.types.json",
    "build:types:downlevel": "yarn downlevel-dts build/types build/types-ts3.8 --to ts3.8",
    "build:watch": "run-p build:transpile:watch build:types:watch",
    "build:dev:watch": "yarn build:watch",
    "build:transpile:watch": "rollup -c rollup.npm.config.mjs --watch",
    "build:types:watch": "tsc -p tsconfig.types.json --watch",
    "build:tarball": "ts-node ../../scripts/prepack.ts && npm pack ./build",
    "circularDepCheck": "madge --circular src/index.ts",
    "clean": "rimraf build coverage sentry-core-*.tgz",
    "fix": "eslint . --format stylish --fix",
    "lint": "eslint . --format stylish",
    "test": "jest",
    "test:watch": "jest --watch",
    "version": "node ../../scripts/versionbump.js src/version.ts",
    "yalc:publish": "ts-node ../../scripts/prepack.ts && yalc publish build --push --sig"
  },
  "volta": {
    "extends": "../../package.json"
  },
  "madge": {
    "detectiveOptions": {
      "ts": {
        "skipTypeImports": true
      }
    }
  },
  "sideEffects": false
}
