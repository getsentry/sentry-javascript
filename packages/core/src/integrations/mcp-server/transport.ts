/**
 * Transport layer instrumentation for MCP server
 *
 * Handles message interception and response correlation.
 * @see https://modelcontextprotocol.io/specification/2025-06-18/basic/transports
 */

import { getIsolationScope, withIsolationScope } from '../../currentScopes';
import { startInactiveSpan, withActiveSpan } from '../../tracing';
import { fill } from '../../utils/object';
import {
  extractSessionDataFromInitializeRequest,
  extractSessionDataFromInitializeResponse,
} from './attributeExtraction';
import { cleanupPendingSpansForTransport, completeSpanWithResults, storeSpanForRequest } from './correlation';
import { captureError } from './errorCapture';
import {
  cleanupSessionDataForTransport,
  storeSessionDataForTransport,
  updateSessionDataForTransport,
} from './sessionManagement';
import { buildMcpServerSpanConfig, createMcpNotificationSpan, createMcpOutgoingNotificationSpan } from './spans';
import type { ExtraHandlerData, MCPTransport } from './types';
import { isJsonRpcNotification, isJsonRpcRequest, isJsonRpcResponse } from './validation';

/**
 * Wraps transport.onmessage to create spans for incoming messages.
 * For "initialize" requests, extracts and stores client info and protocol version
 * in the session data for the transport.
 * @param transport - MCP transport instance to wrap
 */
export function wrapTransportOnMessage(transport: MCPTransport): void {
  if (transport.onmessage) {
    fill(transport, 'onmessage', originalOnMessage => {
      return function (this: MCPTransport, message: unknown, extra?: unknown) {
        if (isJsonRpcRequest(message)) {
          if (message.method === 'initialize') {
            try {
              const sessionData = extractSessionDataFromInitializeRequest(message);
              storeSessionDataForTransport(this, sessionData);
            } catch {
              // noop
            }
          }

          const isolationScope = getIsolationScope().clone();

          return withIsolationScope(isolationScope, () => {
            const spanConfig = buildMcpServerSpanConfig(message, this, extra as ExtraHandlerData);
            const span = startInactiveSpan(spanConfig);

            storeSpanForRequest(this, message.id, span, message.method);

            return withActiveSpan(span, () => {
              return (originalOnMessage as (...args: unknown[]) => unknown).call(this, message, extra);
            });
          });
        }

        if (isJsonRpcNotification(message)) {
          return createMcpNotificationSpan(message, this, extra as ExtraHandlerData, () => {
            return (originalOnMessage as (...args: unknown[]) => unknown).call(this, message, extra);
          });
        }

        return (originalOnMessage as (...args: unknown[]) => unknown).call(this, message, extra);
      };
    });
  }
}

/**
 * Wraps transport.send to handle outgoing messages and response correlation.
 * For "initialize" responses, extracts and stores protocol version and server info
 * in the session data for the transport.
 * @param transport - MCP transport instance to wrap
 */
export function wrapTransportSend(transport: MCPTransport): void {
  if (transport.send) {
    fill(transport, 'send', originalSend => {
      return async function (this: MCPTransport, message: unknown) {
        if (isJsonRpcNotification(message)) {
          return createMcpOutgoingNotificationSpan(message, this, () => {
            return (originalSend as (...args: unknown[]) => unknown).call(this, message);
          });
        }

        if (isJsonRpcResponse(message)) {
          if (message.id !== null && message.id !== undefined) {
            if (message.error) {
              captureJsonRpcErrorResponse(message.error);
            }

            if (message.result && typeof message.result === 'object') {
              const result = message.result as Record<string, unknown>;
              if (result.protocolVersion || result.serverInfo) {
                try {
                  const serverData = extractSessionDataFromInitializeResponse(message.result);
                  updateSessionDataForTransport(this, serverData);
                } catch {
                  // noop
                }
              }
            }

            completeSpanWithResults(this, message.id, message.result);
          }
        }

        return (originalSend as (...args: unknown[]) => unknown).call(this, message);
      };
    });
  }
}

/**
 * Wraps transport.onclose to clean up pending spans for this transport only
 * @param transport - MCP transport instance to wrap
 */
export function wrapTransportOnClose(transport: MCPTransport): void {
  if (transport.onclose) {
    fill(transport, 'onclose', originalOnClose => {
      return function (this: MCPTransport, ...args: unknown[]) {
        cleanupPendingSpansForTransport(this);
        cleanupSessionDataForTransport(this);
        return (originalOnClose as (...args: unknown[]) => unknown).call(this, ...args);
      };
    });
  }
}

/**
 * Wraps transport error handlers to capture connection errors
 * @param transport - MCP transport instance to wrap
 */
export function wrapTransportError(transport: MCPTransport): void {
  if (transport.onerror) {
    fill(transport, 'onerror', (originalOnError: (error: Error) => void) => {
      return function (this: MCPTransport, error: Error) {
        captureTransportError(error);
        return originalOnError.call(this, error);
      };
    });
  }
}

/**
 * Captures JSON-RPC error responses for server-side errors.
 * @see https://www.jsonrpc.org/specification#error_object
 * @internal
 * @param errorResponse - JSON-RPC error response
 */
function captureJsonRpcErrorResponse(errorResponse: unknown): void {
  try {
    if (errorResponse && typeof errorResponse === 'object' && 'code' in errorResponse && 'message' in errorResponse) {
      const jsonRpcError = errorResponse as { code: number; message: string; data?: unknown };

      const isServerError =
        jsonRpcError.code === -32603 || (jsonRpcError.code >= -32099 && jsonRpcError.code <= -32000);

      if (isServerError) {
        const error = new Error(jsonRpcError.message);
        error.name = `JsonRpcError_${jsonRpcError.code}`;

        captureError(error, 'protocol');
      }
    }
  } catch {
    // noop
  }
}

/**
 * Captures transport connection errors
 * @internal
 * @param error - Transport error
 */
function captureTransportError(error: Error): void {
  try {
    captureError(error, 'transport');
  } catch {
    // noop
  }
}
