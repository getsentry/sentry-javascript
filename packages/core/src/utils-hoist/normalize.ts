import type { Primitive } from '../types-hoist';

import { isSyntheticEvent, isVueViewModel } from './is';
import type { MemoFunc } from './memo';
import { memoBuilder } from './memo';
import { convertToPlainObject } from './object';
import { getFunctionName } from './stacktrace';

type Prototype = { constructor: (...args: unknown[]) => unknown };
// This is a hack to placate TS, relying on the fact that technically, arrays are objects with integer keys. Normally we
// think of those keys as actual numbers, but `arr['0']` turns out to work just as well as `arr[0]`, and doing it this
// way lets us use a single type in the places where behave as if we are only dealing with objects, even if some of them
// might be arrays.
type ObjOrArray<T> = { [key: string]: T };

/**
 * Recursively normalizes the given object.
 *
 * - Creates a copy to prevent original input mutation
 * - Skips non-enumerable properties
 * - When stringifying, calls `toJSON` if implemented
 * - Removes circular references
 * - Translates non-serializable values (`undefined`/`NaN`/functions) to serializable format
 * - Translates known global objects/classes to a string representations
 * - Takes care of `Error` object serialization
 * - Optionally limits depth of final output
 * - Optionally limits number of properties/elements included in any single object/array
 *
 * @param input The object to be normalized.
 * @param depth The max depth to which to normalize the object. (Anything deeper stringified whole.)
 * @param maxProperties The max number of elements or properties to be included in any single array or
 * object in the normalized output.
 * @returns A normalized version of the object, or `"**non-serializable**"` if any errors are thrown during normalization.
 */
// eslint-disable-next-line @typescript-eslint/no-explicit-any
export function normalize(input: unknown, depth: number = 100, maxProperties: number = +Infinity): any {
  try {
    // since we're at the outermost level, we don't provide a key
    return visit('', input, depth, maxProperties);
  } catch (err) {
    return { ERROR: `**non-serializable** (${err})` };
  }
}

/** JSDoc */
export function normalizeToSize<T>(
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  object: { [key: string]: any },
  // Default Node.js REPL depth
  depth: number = 3,
  // 100kB, as 200kB is max payload size, so half sounds reasonable
  maxSize: number = 100 * 1024,
): T {
  const normalized = normalize(object, depth);

  if (jsonSize(normalized) > maxSize) {
    return normalizeToSize(object, depth - 1, maxSize);
  }

  return normalized as T;
}

/**
 * Visits a node to perform normalization on it
 *
 * @param key The key corresponding to the given node
 * @param value The node to be visited
 * @param depth Optional number indicating the maximum recursion depth
 * @param maxProperties Optional maximum number of properties/elements included in any single object/array
 * @param memo Optional Memo class handling decycling
 */
function visit(
  key: string,
  value: unknown,
  depth: number = +Infinity,
  maxProperties: number = +Infinity,
  // eslint-disable-next-line deprecation/deprecation
  memo: MemoFunc = memoBuilder(),
): Primitive | ObjOrArray<unknown> {
  const [memoize, unmemoize] = memo;

  // Get the simple cases out of the way first
  if (
    value == null || // this matches null and undefined -> eqeq not eqeqeq
    ['boolean', 'string'].includes(typeof value) ||
    (typeof value === 'number' && Number.isFinite(value))
  ) {
    return value as Primitive;
  }

  const stringified = stringifyValue(key, value);

  // Anything we could potentially dig into more (objects or arrays) will have come back as `"[object XXXX]"`.
  // Everything else will have already been serialized, so if we don't see that pattern, we're done.
  if (!stringified.startsWith('[object ')) {
    return stringified;
  }

  // From here on, we can assert that `value` is either an object or an array.

  // Do not normalize objects that we know have already been normalized. As a general rule, the
  // "__sentry_skip_normalization__" property should only be used sparingly and only should only be set on objects that
  // have already been normalized.
  if ((value as ObjOrArray<unknown>)['__sentry_skip_normalization__']) {
    return value as ObjOrArray<unknown>;
  }

  // We can set `__sentry_override_normalization_depth__` on an object to ensure that from there
  // We keep a certain amount of depth.
  // This should be used sparingly, e.g. we use it for the redux integration to ensure we get a certain amount of state.
  const remainingDepth =
    typeof (value as ObjOrArray<unknown>)['__sentry_override_normalization_depth__'] === 'number'
      ? ((value as ObjOrArray<unknown>)['__sentry_override_normalization_depth__'] as number)
      : depth;

  // We're also done if we've reached the max depth
  if (remainingDepth === 0) {
    // At this point we know `serialized` is a string of the form `"[object XXXX]"`. Clean it up so it's just `"[XXXX]"`.
    return stringified.replace('object ', '');
  }

  // If we've already visited this branch, bail out, as it's circular reference. If not, note that we're seeing it now.
  if (memoize(value)) {
    return '[Circular ~]';
  }

  // If the value has a `toJSON` method, we call it to extract more information
  const valueWithToJSON = value as unknown & { toJSON?: () => unknown };
  if (valueWithToJSON && typeof valueWithToJSON.toJSON === 'function') {
    try {
      const jsonValue = valueWithToJSON.toJSON();
      // We need to normalize the return value of `.toJSON()` in case it has circular references
      return visit('', jsonValue, remainingDepth - 1, maxProperties, memo);
    } catch (err) {
      // pass (The built-in `toJSON` failed, but we can still try to do it ourselves)
    }
  }

  // At this point we know we either have an object or an array, we haven't seen it before, and we're going to recurse
  // because we haven't yet reached the max depth. Create an accumulator to hold the results of visiting each
  // property/entry, and keep track of the number of items we add to it.
  const normalized = (Array.isArray(value) ? [] : {}) as ObjOrArray<unknown>;
  let numAdded = 0;

  // Before we begin, convert`Error` and`Event` instances into plain objects, since some of each of their relevant
  // properties are non-enumerable and otherwise would get missed.
  const visitable = convertToPlainObject(value as ObjOrArray<unknown>);

  for (const visitKey in visitable) {
    // Avoid iterating over fields in the prototype if they've somehow been exposed to enumeration.
    if (!Object.prototype.hasOwnProperty.call(visitable, visitKey)) {
      continue;
    }

    if (numAdded >= maxProperties) {
      normalized[visitKey] = '[MaxProperties ~]';
      break;
    }

    // Recursively visit all the child nodes
    const visitValue = visitable[visitKey];
    normalized[visitKey] = visit(visitKey, visitValue, remainingDepth - 1, maxProperties, memo);

    numAdded++;
  }

  // Once we've visited all the branches, remove the parent from memo storage
  unmemoize(value);

  // Return accumulated values
  return normalized;
}

/* eslint-disable complexity */
/**
 * Stringify the given value. Handles various known special values and types.
 *
 * Not meant to be used on simple primitives which already have a string representation, as it will, for example, turn
 * the number 1231 into "[Object Number]", nor on `null`, as it will throw.
 *
 * @param value The value to stringify
 * @returns A stringified representation of the given value
 */
function stringifyValue(
  key: unknown,
  // this type is a tiny bit of a cheat, since this function does handle NaN (which is technically a number), but for
  // our internal use, it'll do
  value: Exclude<unknown, string | number | boolean | null>,
): string {
  try {
    if (key === 'domain' && value && typeof value === 'object' && (value as { _events: unknown })._events) {
      return '[Domain]';
    }

    if (key === 'domainEmitter') {
      return '[DomainEmitter]';
    }

    // It's safe to use `global`, `window`, and `document` here in this manner, as we are asserting using `typeof` first
    // which won't throw if they are not present.

    if (typeof global !== 'undefined' && value === global) {
      return '[Global]';
    }

    // eslint-disable-next-line no-restricted-globals
    if (typeof window !== 'undefined' && value === window) {
      return '[Window]';
    }

    // eslint-disable-next-line no-restricted-globals
    if (typeof document !== 'undefined' && value === document) {
      return '[Document]';
    }

    if (isVueViewModel(value)) {
      return '[VueViewModel]';
    }

    // React's SyntheticEvent thingy
    if (isSyntheticEvent(value)) {
      return '[SyntheticEvent]';
    }

    if (typeof value === 'number' && !Number.isFinite(value)) {
      return `[${value}]`;
    }

    if (typeof value === 'function') {
      return `[Function: ${getFunctionName(value)}]`;
    }

    if (typeof value === 'symbol') {
      return `[${String(value)}]`;
    }

    // stringified BigInts are indistinguishable from regular numbers, so we need to label them to avoid confusion
    if (typeof value === 'bigint') {
      return `[BigInt: ${String(value)}]`;
    }

    // Now that we've knocked out all the special cases and the primitives, all we have left are objects. Simply casting
    // them to strings means that instances of classes which haven't defined their `toStringTag` will just come out as
    // `"[object Object]"`. If we instead look at the constructor's name (which is the same as the name of the class),
    // we can make sure that only plain objects come out that way.
    const objName = getConstructorName(value);

    // Handle HTML Elements
    if (/^HTML(\w*)Element$/.test(objName)) {
      return `[HTMLElement: ${objName}]`;
    }

    return `[object ${objName}]`;
  } catch (err) {
    return `**non-serializable** (${err})`;
  }
}
/* eslint-enable complexity */

function getConstructorName(value: unknown): string {
  const prototype: Prototype | null = Object.getPrototypeOf(value);

  return prototype ? prototype.constructor.name : 'null prototype';
}

/** Calculates bytes size of input string */
function utf8Length(value: string): number {
  // eslint-disable-next-line no-bitwise
  return ~-encodeURI(value).split(/%..|./).length;
}

/** Calculates bytes size of input object */
// eslint-disable-next-line @typescript-eslint/no-explicit-any
function jsonSize(value: any): number {
  return utf8Length(JSON.stringify(value));
}

/**
 * Normalizes URLs in exceptions and stacktraces to a base path so Sentry can fingerprint
 * across platforms and working directory.
 *
 * @param url The URL to be normalized.
 * @param basePath The application base path.
 * @returns The normalized URL.
 */
export function normalizeUrlToBase(url: string, basePath: string): string {
  const escapedBase = basePath
    // Backslash to forward
    .replace(/\\/g, '/')
    // Escape RegExp special characters
    .replace(/[|\\{}()[\]^$+*?.]/g, '\\$&');

  let newUrl = url;
  try {
    newUrl = decodeURI(url);
  } catch (_Oo) {
    // Sometime this breaks
  }
  return (
    newUrl
      .replace(/\\/g, '/')
      .replace(/webpack:\/?/g, '') // Remove intermediate base path
      // eslint-disable-next-line @sentry-internal/sdk/no-regexp-constructor
      .replace(new RegExp(`(file://)?/*${escapedBase}/*`, 'ig'), 'app:///')
  );
}
