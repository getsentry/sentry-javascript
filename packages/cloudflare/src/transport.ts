import type { BaseTransportOptions, Transport, TransportMakeRequestResponse, TransportRequest } from '@sentry/core';
import { SentryError, createTransport, suppressTracing } from '@sentry/core';

export interface CloudflareTransportOptions extends BaseTransportOptions {
  /** Fetch API init parameters. */
  fetchOptions?: RequestInit;
  /** Custom headers for the transport. */
  headers?: { [key: string]: string };
}

const DEFAULT_TRANSPORT_BUFFER_SIZE = 30;

/**
 * This is a modified promise buffer that collects tasks until drain is called.
 * We need this in the edge runtime because edge function invocations may not share I/O objects, like fetch requests
 * and responses, and the normal PromiseBuffer inherently buffers stuff inbetween incoming requests.
 *
 * A limitation we need to be aware of is that DEFAULT_TRANSPORT_BUFFER_SIZE is the maximum amount of payloads the
 * SDK can send for a given edge function invocation.
 */
export class IsolatedPromiseBuffer {
  // We just have this field because the promise buffer interface requires it.
  // If we ever remove it from the interface we should also remove it here.
  public $: Array<PromiseLike<TransportMakeRequestResponse>>;

  private _taskProducers: (() => PromiseLike<TransportMakeRequestResponse>)[];

  private readonly _bufferSize: number;

  public constructor(_bufferSize = DEFAULT_TRANSPORT_BUFFER_SIZE) {
    this.$ = [];
    this._taskProducers = [];
    this._bufferSize = _bufferSize;
  }

  /**
   * @inheritdoc
   */
  public add(taskProducer: () => PromiseLike<TransportMakeRequestResponse>): PromiseLike<TransportMakeRequestResponse> {
    if (this._taskProducers.length >= this._bufferSize) {
      return Promise.reject(new SentryError('Not adding Promise because buffer limit was reached.'));
    }

    this._taskProducers.push(taskProducer);
    return Promise.resolve({});
  }

  /**
   * @inheritdoc
   */
  public drain(timeout?: number): PromiseLike<boolean> {
    const oldTaskProducers = [...this._taskProducers];
    this._taskProducers = [];

    return new Promise(resolve => {
      const timer = setTimeout(() => {
        if (timeout && timeout > 0) {
          resolve(false);
        }
      }, timeout);

      // This cannot reject
      // eslint-disable-next-line @typescript-eslint/no-floating-promises
      Promise.all(
        oldTaskProducers.map(taskProducer =>
          taskProducer().then(null, () => {
            // catch all failed requests
          }),
        ),
      ).then(() => {
        // resolve to true if all fetch requests settled
        clearTimeout(timer);
        resolve(true);
      });
    });
  }
}

/**
 * Creates a Transport that uses the native fetch API to send events to Sentry.
 */
export function makeCloudflareTransport(options: CloudflareTransportOptions): Transport {
  function makeRequest(request: TransportRequest): PromiseLike<TransportMakeRequestResponse> {
    const requestOptions: RequestInit = {
      body: request.body,
      method: 'POST',
      headers: options.headers,
      ...options.fetchOptions,
    };

    return suppressTracing(() => {
      return fetch(options.url, requestOptions).then(response => {
        return {
          statusCode: response.status,
          headers: {
            'x-sentry-rate-limits': response.headers.get('X-Sentry-Rate-Limits'),
            'retry-after': response.headers.get('Retry-After'),
          },
        };
      });
    });
  }

  return createTransport(options, makeRequest, new IsolatedPromiseBuffer(options.bufferSize));
}
