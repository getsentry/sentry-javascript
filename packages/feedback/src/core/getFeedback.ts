import { getClient } from '@sentry/core';
import type { buildFeedbackIntegration } from './integration';

type FeedbackIntegration = ReturnType<typeof buildFeedbackIntegration>;

/**
 * This is a small utility to get a type-safe instance of the Feedback integration.
 */
export function getFeedback(): ReturnType<FeedbackIntegration> | undefined {
  const client = getClient();
  return client && client.getIntegrationByName<ReturnType<FeedbackIntegration>>('Feedback');
}
