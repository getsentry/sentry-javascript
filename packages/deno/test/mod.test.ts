import { assertEquals } from 'https://deno.land/std@0.202.0/assert/assert_equals.ts';
import { assertSnapshot } from 'https://deno.land/std@0.202.0/testing/snapshot.ts';

import type { Event } from '@sentry/core';
import { createStackParser, nodeStackLineParser } from '@sentry/core';
import { DenoClient, getCurrentScope, getDefaultIntegrations } from '../build/esm/index.js';

import { getNormalizedEvent } from './normalize.ts';
import { makeTestTransport } from './transport.ts';

function getTestClient(callback: (event?: Event) => void): DenoClient {
  const client = new DenoClient({
    dsn: 'https://233a45e5efe34c47a3536797ce15dafa@nothing.here/5650507',
    debug: true,
    integrations: getDefaultIntegrations({}),
    stackParser: createStackParser(nodeStackLineParser()),
    transport: makeTestTransport(envelope => {
      callback(getNormalizedEvent(envelope));
    }),
  });

  client.init();
  getCurrentScope().setClient(client);

  return client;
}

function delay(time: number): Promise<void> {
  return new Promise(resolve => {
    setTimeout(resolve, time);
  });
}

Deno.test('captureException', async t => {
  let ev: Event | undefined;
  const client = getTestClient(event => {
    ev = event;
  });

  function something() {
    return new Error('Some unhandled error');
  }

  client.captureException(something());

  await delay(200);
  await assertSnapshot(t, ev);
});

Deno.test('captureMessage', async t => {
  let ev: Event | undefined;
  const client = getTestClient(event => {
    ev = event;
  });

  client.captureMessage('Some error message');

  await delay(200);
  await assertSnapshot(t, ev);
});

Deno.test('captureMessage twice', async t => {
  let ev: Event | undefined;
  const client = getTestClient(event => {
    ev = event;
  });

  client.captureMessage('Some error message');

  await delay(200);
  await assertSnapshot(t, ev);

  client.captureMessage('Another error message');

  await delay(200);
  await assertSnapshot(t, ev);
});

Deno.test('App runs without errors', async _ => {
  const cmd = new Deno.Command('deno', {
    args: ['run', '--allow-net=some-domain.com', './test/example.ts'],
    stdout: 'piped',
    stderr: 'piped',
  });

  const output = await cmd.output();
  assertEquals(output.success, true);

  const td = new TextDecoder();
  const outString = td.decode(output.stdout);
  const errString = td.decode(output.stderr);
  assertEquals(outString, 'App has started\n');
  assertEquals(errString, '');
});
