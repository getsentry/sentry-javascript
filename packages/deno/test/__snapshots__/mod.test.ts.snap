export const snapshot = {};

snapshot[`captureException 1`] = `
{
  contexts: {
    app: {
      app_start_time: "{{time}}",
    },
    device: {
      arch: "{{arch}}",
      processor_count: 0,
    },
    os: {
      name: "{{platform}}",
      version: "{{version}}",
    },
    runtime: {
      name: "deno",
      version: "{{version}}",
    },
    trace: {
      span_id: "{{id}}",
      trace_id: "{{id}}",
    },
    typescript: {
      name: "TypeScript",
      version: "{{version}}",
    },
    v8: {
      name: "v8",
      version: "{{version}}",
    },
  },
  environment: "production",
  event_id: "{{id}}",
  exception: {
    values: [
      {
        mechanism: {
          handled: true,
          type: "generic",
        },
        stacktrace: {
          frames: [
            {
              colno: 27,
              context_line: "  client.captureException(something());",
              filename: "app:///test/mod.test.ts",
              function: "?",
              in_app: true,
              lineno: 44,
              post_context: [
                "",
                "  await delay(200);",
                "  await assertSnapshot(t, ev);",
                "});",
                "",
                "Deno.test('captureMessage', async t => {",
                "  let ev: Event | undefined;",
              ],
              pre_context: [
                "    ev = event;",
                "  });",
                "",
                "  function something() {",
                "    return new Error('Some unhandled error');",
                "  }",
                "",
              ],
            },
            {
              colno: 12,
              context_line: "    return new Error('Some unhandled error');",
              filename: "app:///test/mod.test.ts",
              function: "something",
              in_app: true,
              lineno: 41,
              post_context: [
                "  }",
                "",
                "  client.captureException(something());",
                "",
                "  await delay(200);",
                "  await assertSnapshot(t, ev);",
                "});",
              ],
              pre_context: [
                "Deno.test('captureException', async t => {",
                "  let ev: Event | undefined;",
                "  const client = getTestClient(event => {",
                "    ev = event;",
                "  });",
                "",
                "  function something() {",
              ],
            },
          ],
        },
        type: "Error",
        value: "Some unhandled error",
      },
    ],
  },
  level: "error",
  platform: "javascript",
  sdk: {
    integrations: [
      "InboundFilters",
      "FunctionToString",
      "LinkedErrors",
      "Dedupe",
      "Breadcrumbs",
      "DenoContext",
      "ContextLines",
      "NormalizePaths",
      "GlobalHandlers",
    ],
    name: "sentry.javascript.deno",
    packages: [
      {
        name: "denoland:sentry",
        version: "{{version}}",
      },
    ],
    version: "{{version}}",
  },
  server_name: "{{server}}",
  timestamp: 0,
}
`;

snapshot[`captureMessage 1`] = `
{
  breadcrumbs: [
    {
      category: "sentry.event",
      event_id: "{{id}}",
      level: "error",
      message: "Error: Some unhandled error",
      timestamp: 0,
    },
  ],
  contexts: {
    app: {
      app_start_time: "{{time}}",
    },
    device: {
      arch: "{{arch}}",
      processor_count: 0,
    },
    os: {
      name: "{{platform}}",
      version: "{{version}}",
    },
    runtime: {
      name: "deno",
      version: "{{version}}",
    },
    trace: {
      span_id: "{{id}}",
      trace_id: "{{id}}",
    },
    typescript: {
      name: "TypeScript",
      version: "{{version}}",
    },
    v8: {
      name: "v8",
      version: "{{version}}",
    },
  },
  environment: "production",
  event_id: "{{id}}",
  level: "info",
  message: "Some error message",
  platform: "javascript",
  sdk: {
    integrations: [
      "InboundFilters",
      "FunctionToString",
      "LinkedErrors",
      "Dedupe",
      "Breadcrumbs",
      "DenoContext",
      "ContextLines",
      "NormalizePaths",
      "GlobalHandlers",
    ],
    name: "sentry.javascript.deno",
    packages: [
      {
        name: "denoland:sentry",
        version: "{{version}}",
      },
    ],
    version: "{{version}}",
  },
  server_name: "{{server}}",
  timestamp: 0,
}
`;

snapshot[`captureMessage twice 1`] = `
{
  breadcrumbs: [
    {
      category: "sentry.event",
      event_id: "{{id}}",
      level: "error",
      message: "Error: Some unhandled error",
      timestamp: 0,
    },
    {
      category: "sentry.event",
      event_id: "{{id}}",
      level: "info",
      message: "Some error message",
      timestamp: 0,
    },
  ],
  contexts: {
    app: {
      app_start_time: "{{time}}",
    },
    device: {
      arch: "{{arch}}",
      processor_count: 0,
    },
    os: {
      name: "{{platform}}",
      version: "{{version}}",
    },
    runtime: {
      name: "deno",
      version: "{{version}}",
    },
    trace: {
      span_id: "{{id}}",
      trace_id: "{{id}}",
    },
    typescript: {
      name: "TypeScript",
      version: "{{version}}",
    },
    v8: {
      name: "v8",
      version: "{{version}}",
    },
  },
  environment: "production",
  event_id: "{{id}}",
  level: "info",
  message: "Some error message",
  platform: "javascript",
  sdk: {
    integrations: [
      "InboundFilters",
      "FunctionToString",
      "LinkedErrors",
      "Dedupe",
      "Breadcrumbs",
      "DenoContext",
      "ContextLines",
      "NormalizePaths",
      "GlobalHandlers",
    ],
    name: "sentry.javascript.deno",
    packages: [
      {
        name: "denoland:sentry",
        version: "{{version}}",
      },
    ],
    version: "{{version}}",
  },
  server_name: "{{server}}",
  timestamp: 0,
}
`;

snapshot[`captureMessage twice 2`] = `
{
  breadcrumbs: [
    {
      category: "sentry.event",
      event_id: "{{id}}",
      level: "error",
      message: "Error: Some unhandled error",
      timestamp: 0,
    },
    {
      category: "sentry.event",
      event_id: "{{id}}",
      level: "info",
      message: "Some error message",
      timestamp: 0,
    },
    {
      category: "sentry.event",
      event_id: "{{id}}",
      level: "info",
      message: "Some error message",
      timestamp: 0,
    },
  ],
  contexts: {
    app: {
      app_start_time: "{{time}}",
    },
    device: {
      arch: "{{arch}}",
      processor_count: 0,
    },
    os: {
      name: "{{platform}}",
      version: "{{version}}",
    },
    runtime: {
      name: "deno",
      version: "{{version}}",
    },
    trace: {
      span_id: "{{id}}",
      trace_id: "{{id}}",
    },
    typescript: {
      name: "TypeScript",
      version: "{{version}}",
    },
    v8: {
      name: "v8",
      version: "{{version}}",
    },
  },
  environment: "production",
  event_id: "{{id}}",
  level: "info",
  message: "Another error message",
  platform: "javascript",
  sdk: {
    integrations: [
      "InboundFilters",
      "FunctionToString",
      "LinkedErrors",
      "Dedupe",
      "Breadcrumbs",
      "DenoContext",
      "ContextLines",
      "NormalizePaths",
      "GlobalHandlers",
    ],
    name: "sentry.javascript.deno",
    packages: [
      {
        name: "denoland:sentry",
        version: "{{version}}",
      },
    ],
    version: "{{version}}",
  },
  server_name: "{{server}}",
  timestamp: 0,
}
`;
